### PR TITLE
make `rationalize` default to use `Int` on `Irrational` (and `Real`)

### PR DESCRIPTION
**Pull request by MasonProtter**
Alternative to https://github.com/JuliaLang/julia/pull/46625

This would make `rationalize(::AbstractIrrational)` consistent with `rationalize(::AbstractFloat)`

What would be the recommended way to test this? Do we want to test that say `rationalize(pi)` produces **exactly** `2646693125139304345//842468587426513207`, or just approximately? Does this even need a test case?

### CODE REVIEW

**Comment by udohjeremiah** [comment]
@MasonProtter I don't think its utterly useful making `π` or other `Irrational` values default to `Int` on the `rationalize` function.

While defaulting to `Int` might be a nice solution to make it consistent with `rationalize(::AbstractFloat)`, but then it introduces a whole new topic.

A few of them are:
- `Int` is NOT machine-independent. And yes most computers are `64-bit`, but such assumptions won't make sense when a user is using a `32-bit` machine. Two users using different machines will get slightly different behaviours.
- It's stated all over the docs that `Irrational` values will always default to `Float64` when used in any arithmetic operations or implicitly converted. Making `Irrational` values default to `Int` on ONLY the`rationalize` would need an explanation and then users might expect same behaviours in other areas of Julia, which of course will throw an error since its not the standard.

There are few other reasons but I'd stop here for now and probably see what others think of this.

**Comment by MasonProtter** [comment]
> * `Int` is NOT machine-independent. And yes most computers are `64-bit`, but such assumptions won't make sense when a user is using a `32-bit` machine. Two users using different machines will get slightly different behaviours.

This is standard and consistent with how we treat anything in Julia that we convert to `Int`. Users of 32 bit computers are warned about this and expected to know it. This complaint you are making applies equally well to any other application of `rationalize(::AbstractFloat)`. 

> * It's stated all over the docs that `Irrational` values will always default to `Float64` when used in any arithmetic operations or implicitly converted. Making `Irrational` values default to `Int` on ONLY the`rationalize` would need an explanation and then users might expect same behaviours in other areas of Julia, which of course will throw an error since its not the standard.

That's not at all inconsistent. `rationalize` is an explicit approximate conversion, *not* and implicit one, nor is it a standard arithmetic operation.

**Comment by udohjeremiah** [comment]
Before I add my point, I'd speak on the pronoun you used:
> This is standard and consistent with how we treat anything in Julia that we convert to `Int`.

Anyone who pulls a request or comments on any issue in the julia repo, in a sense means good for Julia; so I request you stop using the pronoun "we" - denoting the one you're referring to is outside the Julia community in some way, and use collective nouns like "the community" - referring to the whole community at large. This would promote better open source atmosphere.

So
> This is standard and consistent with how "the community" treats anything in Julia that the language convert to `Int`.

is better and conducive than:

> This is standard and consistent with how we treat anything in Julia that we convert to `Int`.


Speaking on this:
> That's not at all inconsistent. `rationalize` is an explicit approximate conversion, not an implicit one, nor is it a standard arithmetic operation.

If what you said there should hold true, then shouldn't we have a method for `trunc(::Type{T}, ::Irrational) where T<:Union{Signed, Unsigned}` defined as well.

```julia
julia> trunc(2.9)          # this works
2.0

julia> trunc(Int, 2.9)    # this works
2

julia> trunc(π)             # this works also
3.0

julia> trunc(Int, π)    # if rationalize(π) should work, why should trunc(Int, π) fail
ERROR: MethodError: no method matching trunc(::Type{Int64}, ::Irrational{:π})
```

I could go simpler to this even:
```julia
julia> Int(π)
ERROR: MethodError: no method matching Int64(::Irrational{:π})
```

One might argue that `Int(π)` would never work in the first place as it throws an `InexactError`, but of course the error message reads `MethodError`, meaning its not even considered VALID.

My point is: `Irrational` values should be best left to floating-types only, as they are understandable from that perspective. Trying to treat every corner cases in functions where `Irrational` values are implicitly converted to `Int` would open up a new wave of applications for `Irrational` values to work in functions types as I highlighted above. Conversion for `Irrational` values to any type other than floating-points should be EXPLICIT and "makes technical sense"; anything other than that would bring in new confusions.

Anyone would think, if `Irrational` values work on this instance:
```julia
julia> rationalize(π)
2646693125139304345//842468587426513207
```

how come it fails in this:
```julia
julia> trunc(Int, π)
ERROR: MethodError: no method matching trunc(::Type{Int64}, ::Irrational{:π})
```

`trunc` says:
```julia
help?> trunc
search: trunc truncate unsafe_trunc setrounding AbstractUnitRange

  trunc(T, x) converts the result to type T, throwing an InexactError if the
  value is not representable.
```

Well if your solution is accepted for `rationalize(π)` to work, then I think `trunc(Int, π)` should work as well. Since it doesn't make technical sense for `rationalize` to implicitly convert to `Int` but then `trunc` throws an `InexactError` for `Irrational` values.

Moreover on what you said, I find `trunc` being more "EXPLICIT CONVERSION" than `rationalize` does, yet it fails when mangling `Int` with `Irrational` values.

**Comment by MasonProtter** [comment]
> If what you said there should hold true, then shouldn't we have a method for `trunc(::Type{T}, ::Irrational) where T<:Union{Signed, Unsigned}` defined as well.
> 
> ```julia
> julia> trunc(2.9)          # this works
> 2.0
> 
> julia> trunc(Int, 2.9)    # this works
> 2
> 
> julia> trunc(π)             # this works also
> 3.0
> 
> julia> trunc(Int, π)    # if rationalize(π) should work, why should trunc(Int, π) fail
> ERROR: MethodError: no method matching trunc(::Type{Int64}, ::Irrational{:π})
> ```

I don't think this is  relevant, since it's the opposite case to the one you're trying to argue against. i.e. `trunk(pi)` works, but `rationalize(pi)` doesn't. This just seems like a missing method in `trunk` and I see no reason why we shouldn't add a method for `trunk(::Type{T}, ::AbstractIrrational)` if someone asked for it. 

> I could go simpler to this even:
> 
> ```julia
> julia> Int(π)
> ERROR: MethodError: no method matching Int64(::Irrational{:π})
> ```
> 
> One might argue that `Int(π)` would never work in the first place as it throws an `InexactError`, but of course the error message reads `MethodError`, meaning its not even considered VALID.

So what? `Int(1.1)` also throws an error, yet `rationalize(1.1)` is fine. `Int(x::Real)` will only work when `x` can be *exactly* represented as an integer. 

> My point is: `Irrational` values should be best left to floating-types only, as they are understandable from that perspective. Trying to treat every corner cases in functions where `Irrational` values are implicitly converted to `Int` would open up a new wave of applications for `Irrational` values to work in functions types as I highlighted above. Conversion for `Irrational` values to any type other than floating-points should be EXPLICIT and "makes technical sense"; anything other than that would bring in new confusions.

`Irrational` values are symbolic placeholders. Floats cannot exactly represent them, just like rationals cannot exactly represent them. 

The docstring for `rationalize` explicitly says 

> `rationalize([T<:Integer=Int,] x; tol::Real=eps(x))`
>   Approximate floating point number `x` as a `Rational` number with components of the given integer type. The result will differ from `x` by no more than `tol`.

It is fully expected and documented that `rationalize` will approximate a number up to a certain tolerance, and that the default type used for the rationalization is `Int`.

**Comment by MasonProtter** [comment]
> Before I add my point, I'd speak on the pronoun you used:
> 
> > This is standard and consistent with how we treat anything in Julia that we convert to `Int`.
> 
> Anyone who pulls a request or comments on any issue in the julia repo, in a sense means good for Julia; so I request you stop using the pronoun "we" - denoting the one you're referring to is outside the Julia community in some way, and use collective nouns like "the community" - referring to the whole community at large. This would promote better open source atmosphere.
> 
> So
> 
> > This is standard and consistent with how "the community" treats anything in Julia that the language convert to `Int`.
> 
> is better and conducive than:
> 
> > This is standard and consistent with how we treat anything in Julia that we convert to `Int`.
> 
> Speaking on this:
> 
> > That's not at all inconsistent. `rationalize` is an explicit approximate conversion, not an implicit one, nor is it a standard arithmetic operation.

I did not use the word 'we' to try and exclude you. I was including you in that 'we'. But a more accurate thing for me to say would just be to use the word "julia" instead of "the community" because I was talking about existing behaviour in the language itself.

**Comment by udohjeremiah** [comment]
It seems we two are seeing it from a different perspective. I'd love @ViralBShah @StefanKarpinski @JeffBezanson or other contributors to look at it.

Meanwhile, should I add the "triage" bag to this, since this seems more of like it?

**Review by Keno — APPROVED** [review]

**Comment by KristofferC** [comment]
> Do we want to test that say rationalize(pi) produces exactly 2646693125139304345//842468587426513207, or just approximately?

I think that is fine for now. Mostly just to run the code path.

**Comment by StefanKarpinski** [comment]
I'm a bit late here but I agree with @udohjeremiah here: defaulting to machine-dependent `Int` rationals seems bad here. Why? Because it means the value that we return depends on the system, not just its type. That's very different from other uses of `Int` which return the same value but the type depends on the system.

**Comment by MasonProtter** [comment]
But isn't that what we already do all over the place?
```julia
julia> rationalize(rand(Float64))
25154543//405520987

julia> typeof(ans)
Rational{Int64}
```

**Comment by MasonProtter** [comment]
@StefanKarpinski, any updated thoughts for or against? I see this as being in line with all the behaviour for floating point values:
```julia
julia> rationalize(rand(Float32)) |> typeof
Rational{Int64}

julia> rationalize(rand(Float64)) |> typeof
Rational{Int64}

julia> rationalize(rand(BigFloat)) |> typeof
Rational{Int64}
```

File: test/rational.jl [hunk 1]
```diff
@@ -542,6 +542,7 @@ end
              100798//32085
              103993//33102
              312689//99532 ]
+    @test rationalize(pi) == rationalize(BigFloat(pi))
```
Comment (LilithHafner) [inline test/rational.jl:545]:
```suggestion
    @test rationalize(pi) === rationalize(BigFloat(pi))
```
These should have the same type

File: base/irrationals.jl [hunk 2]
```diff
@@ -110,6 +110,8 @@ end
 @assume_effects :total function rationalize(::Type{T}, x::AbstractIrrational; tol::Real=0) where T
     return rationalize(T, big(x), tol=tol)
 end
+rationalize(x::AbstractIrrational; kwargs...) = rationalize(Int, x)
```
Comment (LilithHafner) [inline base/irrationals.jl:113]: For a more bold and general solution, we could change

```julia
rationalize(x::AbstractFloat; kvs...) = rationalize(Int, x; kvs...)
```
to
```julia
rationalize(x::Real; kvs...) = rationalize(Int, x; kvs...)
```
in rational.jl instead

**Review by LilithHafner — COMMENTED** [review]
The docstring of rationalize seems to support this PR, but also currently [promises the impossible](https://github.com/JuliaLang/julia/issues/52335), and also does not document behavior on non-floating point inputts, so I don't take it particularly seriously.

I think it would be nice to have an idea for how to proceed with existing bugs before expanding public functionality.

**Comment by PallHaraldsson** [comment]
I suggest to rather rationalize to Int32 (if not UInt32...), I tested it ok for π and ℯ, but did not test for γ:

>It's stated all over the docs that Irrational values will always default to Float64 when used in any arithmetic operations or implicitly converted. 

I'm not sure that's stated, or actually it's not true:

```
julia> pi + BigFloat(0)  # this is most likely 256-bit accuracy, at least it's not Float64 accuracy:
3.141592653589793238462643383279502884197169399375105820974944592307816406286198

however:
a = rationalize(pi + BigFloat(0))
2646693125139304345//842468587426513207

julia> typeof(numerator(a))
Int64

julia> rationalize(float(pi))
165707065//52746197

julia> rationalize(Int32, float(pi))  # same values different smaller type
165707065//52746197

julia> rationalize(Float32(pi))  # Not same
355//113

julia> pi
π = 3.1415926535897...

julia> float(pi)  # note, doc above show truncated, then ... implying more, and the value before is wrong for rounded. ending in 79... better or was that intentional for some reason?
3.141592653589793

julia> Float32(float(pi))  # that gives seemingly rounded up in binary (also correct for decimal shown), so should trunc(Float32, float(pi)) be allowed, and the redundant round(Float64, pi)?
3.1415927f0
```

**Comment by PallHaraldsson** [comment]
If you rationalize from Float64 to Rational{Int64}, people may _incorrectly_ assume the fraction is the most accurate for that type
all the, but it's not, unless you go from BigFloat.

I still suggest rationalizing to Rational{Int32}. By definition rationalize will give you, well, a rational (and the best you can do for Float32 or lower), so any is an approximation unlike when you rationalize already rational Float64 numbers based on Int64. People are willingly accepting a rational, since asked for, and it's not clear based in Int64 or Int128 or BigInt (with default 256-bit) or whatever is better.

```
julia> pi - 1068966896//340262731 # absolute error based on value from rationalize(Int32, big(pi)), as opposed to rationalize(Int32, float(pi)), or rationalize(Float32(pi))
0.0

julia> pi/(1068966896//340262731)
1.0

julia> big(pi) - 1068966896//340262731
3.070078453798583623603801623225680863823231141797716315424355206651227439588888e-18

julia> big(pi)/(1068966896//340262731)
1.000000000000000000977236323203935201843056515288947918981142826290421628781247

does anyone need more accurate?


julia> big(pi) - (2646693125139304345//842468587426513207)
1.410721535882205084217088334741350691255671024260366177695347682289379611696669e-38

julia> big(pi)/(2646693125139304345//842468587426513207)
1.000000000000000000000000000000000000004490466115236870682183430677486410741887
```

Getting Rational{Int32} gets rid of a difference between 32-bit, and on 64-bit might be stored ion one CPU register, upper and lower half, not sure that will actually happen though. I think that's enough for most, for numbers from even non-irrational.

One fear might be it being too small inducing overflows, but if you involved in calculations with Float32, 64, or larger rationals it disables that threat, and it's not even bad if you keep using the Int32-based Rational type:

There are of course infinite irrational numbers, and arbitrary high in magnitude, but if it helps, all the important ones are between 0 and about 10, o actually maxing at tau = 2pi= 6.28. [Smallest is [Heath-Brown–Moroz constan](https://en.wikipedia.org/wiki/Heath-Brown%E2%80%93Moroz_constant) 0.00131 76411 54853 17810]

Only exceptions from:
https://en.wikipedia.org/wiki/List_of_mathematical_constants

Gelfond's constant π^ℯ (never actually used?) is 23.14069 26327 79269 0057, highest only I know of (then you can double etc.), excerpt for [Ramanujan's constant](https://en.wikipedia.org/wiki/Heegner_number#Almost_integers_and_Ramanujan's_constant) (which is yes very large, but neither used?):

```
julia> 262537412640768743.999999999999250073-262537412640768704//1  # That one to Int64, doesn't even get non-fractional part right:
32.0

julia> 262537412640768743.999999999999250073-262537400750702592//1  #  Looks bad for Rational{Int32}
1.1890066144e10

though relative error not bad:
julia> 262537412640768743.999999999999250073/(262537400750702592//1)
1.0000000452890374
```

**Comment by LilithHafner** [comment]
It would be significantly breaking to change from Int to Int32.

**Comment by PallHaraldsson** [comment]
Why, Int32 is what you get on 32-bit already (or at least will get)? And does anyone rely on Int64-based (on 64-bit) already and then why? It doesn't seem needed. What's good enough for 32-bit is good enough for 64-bit? You are proposing Int which WOULD be different accuracy. Int is Julia's default for mainly addressing reason, not accuracy. Then it seems better to be explicit as with (explicit) choice of a float type.

**Comment by LilithHafner** [comment]
Using Int32 as a default would break `rationalize(float(π), tol=.01) ^ 10` on 64-bit systems. Using Int64 would also be breaking, albeit less so.

**Comment by PallHaraldsson** [comment]
Alternatively, I might be ok with Int64 and 64- and 32-bit too (so the same operations break on both, or succeed on both). But your example seems maybe not too useful.

The help shows `rationalize([T<:Integer=Int,] x; tol::Real=eps(x))` but eps isn't defined on rationals so what would take its place as a default?

Yes, your pi approximation approximation even fails sooner, but do you really need this high a power?

```
julia> (Int32(22)//Int32(7)) ^ 7
ERROR: OverflowError: 10648 * 234256 overflowed for type Int32

then you should be opting into BigInt-based (or Int128? Why not that default? Rationals are slow anyway might as well do that). Otherwise you fail eventually, for Int64-based at:

julia> (22//7)^15
ERROR: OverflowError: 2494357888 * 54875873536 overflowed for type Int64

julia> rationalize(float(π), tol=.00001)  # i.e. 3.1415929203539825
355//113

this closer fails at:
julia> (355//113)^8
ERROR: OverflowError: 15882300625 * 15882300625 overflowed for type Int64

and this one at:
julia> (75948//24175) ^ 4
ERROR: OverflowError: 5768098704 * 5768098704 overflowed for type Int64
```

so what tol would you want the as default, and how high a power do you want possible?

**Comment by LilithHafner** [comment]
For the first two there wasn't much discussion; it just seemed like the obvious thing to do (that's even how it's documented)

For the last one, some points that came up
- PkgEval won't work
- It's not going to change the values people get back as long as they specify a `tol` (and if they don't specify a `tol` then they're already going to overflow right away and also might get wrong answers if there is no rational within `eps(x)`)

File: base/irrationals.jl [hunk 3]
```diff
@@ -110,6 +110,8 @@ end
 @assume_effects :total function rationalize(::Type{T}, x::AbstractIrrational; tol::Real=0) where T
     return rationalize(T, big(x), tol=tol)
 end
+rationalize(x::Real; kwargs...) = rationalize(Int, x; kwargs...)
```
Comment (LilithHafner) [inline base/irrationals.jl:113]: This belongs in `base/rational.jl` (replacing the ::AbstractFloat method)

**Review by LilithHafner — APPROVED** [review]

**Pull request merged**